src/syntax/ast.rs

```diff
@@ -5,101 +5,108 @@ pub struct Resource {
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Comment {
-    pub body: String,
-}
-
-#[derive(Debug, PartialEq)]
-pub struct Section {
-    pub key: Key,
-    pub body: Vec<Entry>,
+pub struct Message {
+    pub id: Identifier,
+    pub value: Option<Pattern>,
+    pub attributes: Option<Vec<Attribute>>,
     pub comment: Option<Comment>,
 }
 
 #[derive(Debug, PartialEq)]
 pub enum Entry {
     Message(Message),
-    Comment(Comment),
     Section(Section),
+    Comment(Comment),
     Junk(JunkEntry),
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Message {
+pub struct Pattern {
+    pub elements: Vec<Expression>,
+    pub quoted: bool,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Expression {
+    Pattern(Pattern),
+    String(String),
+    Number(Number),
+    MessageReference(String),
+    ExternalArgument(String),
+    SelectExpression {
+        exp: Option<Box<Expression>>,
+        variants: Vec<Variant>,
+    },
+    AttributeExpression { id: Identifier, name: Identifier },
+    VariantExpression { id: Identifier, key: VariantKey },
+    CallExpression {
+        callee: Builtin,
+        args: Vec<Argument>,
+    },
+    Expression(Box<Expression>),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Attribute {
     pub id: Identifier,
-    pub value: Option<Pattern>,
-    pub traits: Option<Vec<Member>>,
-    pub comment: Option<Comment>,
+    pub value: Pattern,
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Pattern {
-    pub elements: Vec<PatternElement>,
-    pub quoted: bool,
+pub struct Variant {
+    pub key: VariantKey,
+    pub value: Pattern,
+    pub default: bool,
 }
 
 #[derive(Debug, PartialEq)]
-pub enum PatternElement {
-    Text(String),
-    Placeable { expressions: Vec<Expression> },
+pub enum VariantKey {
+    Key(Keyword),
+    Number(Number),
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Identifier {
-    pub name: String,
+pub enum Argument {
+    Expression(Expression),
+    NamedArgument { name: Identifier, val: ArgValue },
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Key {
+pub enum ArgValue {
+    Number(Number),
+    String(String),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Identifier {
     pub name: String,
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Number {
+pub struct Keyword {
     pub value: String,
```

Comment:
This used to be `name` which is why it's name in ASDL. I don't care either way, should I change it?